bootstrap/src/q.ml

```diff
@@ -0,0 +1,202 @@
+open Rudiments
+
+type 'a t = usize * 'a Stream.t * 'a list * 'a Stream.t
+
+let empty = (0, lazy Stream.Nil, [], lazy Stream.Nil)
+
+let exec (l, f, r, s) =
+  let rec rotate (l, f, r, s) = lazy begin
+    match f, r with
+    | _, [] -> assert false
+    | lazy Stream.Nil, y :: _ -> Lazy.force (Stream.push y s)
+    | lazy (Stream.Cons(elm, f')), y :: y' -> begin
+      let s' = Stream.push y s in
+      let f'' = rotate (l, f', y', s') in
+      Lazy.force (Stream.push elm f'')
+    end
+  end in
+  match s with
+  | lazy (Stream.Cons(_, s')) -> (l, f, r, s')
+  | lazy Stream.Nil -> begin
+    let f' = rotate (l, f, r, s) in
+    (l, f', [], f')
+  end
+
+let length (l, _, _, _) = l
```

Comment:
Thus far my preferred style for top-level declarations like this has been to put the function body on a separate line even if it's simple.  That way we can tell at a glance whether the body is a constant vs a function.  So the `empty` declaration above is fine, but this one needs a line break.  Does that seem like a worthwhile distinction?  I could be convinced either way.

Reply:
Seems like a decent convention.